Data market agency

sweatpotato13 opened the pull request:
# Overview

Add agency logic in `data-market` pallet

This feature is based on Contract. There are two types of Contract.

- Delegate Contract
- Purchase Contract

In detail, see [issue](https://github.com/InfraBlockchain/infrablockchain-substrate/issues/50)

# Changes

* Contracts are stored in storage
```rust
	// The Data Delegate Contracts
	#[pallet::storage]
	#[pallet::getter(fn get_data_delegate_contracts)]
	pub(super) type DataDelegateContracts<T: Config> = StorageMap<
		_,
		Twox64Concat,
		ContractId,
		DataDelegateContractDetail<T::AccountId, BlockNumberFor<T>, AnyText>,
		OptionQuery,
	>;

	// The Data Purchase Contracts
	#[pallet::storage]
	#[pallet::getter(fn get_data_purchase_contracts)]
	pub(super) type DataPurchaseContracts<T: Config> = StorageMap<
		_,
		Twox64Concat,
		ContractId,
		DataPurchaseContractDetail<
			T::AccountId,
			BlockNumberFor<T>,
			<T as pallet_assets::Config>::Balance,
			AnyText,
		>,
		OptionQuery,
	>;
	// The Data Delegate Contract List
	#[pallet::storage]
	#[pallet::getter(fn get_data_delegate_contract_list)]
	pub(super) type DataDelegateContractList<T: Config> =
		StorageMap<_, Twox64Concat, T::AccountId, Vec<ContractId>, ValueQuery>;

	// The Data Purchase Contract List
	#[pallet::storage]
	#[pallet::getter(fn get_data_purchase_contract_list)]
	pub(super) type DataPurchaseContractList<T: Config> =
		StorageMap<_, Twox64Concat, T::AccountId, Vec<ContractId>, ValueQuery>;

// ------- snip -------
```

* Pallet has method for make, sign, terminate contract & execute data trade
```rust 
		#[pallet::call_index(0)]
		pub fn make_delegate_contract

		#[pallet::call_index(1)]
		pub fn sign_delegate_contract

		#[pallet::call_index(2)]
		pub fn make_purchase_contract

		#[pallet::call_index(3)]
		pub fn sign_purchase_contract

		#[pallet::call_index(4)]
		pub fn terminate_delegate_contract

		#[pallet::call_index(5)]
		pub fn terminate_purchase_contract

		#[pallet::call_index(6)]
		pub fn execute_data_trade
```

Closed #50

sweatpotato13 committed:
feat: 🎸 add agency logic in data-market pallet

sweatpotato13 committed:
test: 💍 write data-market pallet test code

sweatpotato13 committed:
test: 💍 add benchmark for data-market pallet

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -31,7 +36,7 @@ pub mod pallet {
 	pub trait Config: frame_system::Config + pallet_assets::Config {
```
현재 instance 분리중인데 tightly-coupled 하게 되면 안됩니다. `fungibles::mutate` 같은 trait 이용해서 `transfer` 같은 기능을 이용해야합니다

sweatpotato13 replied:
Resolved 561dd9ee5b80fd34a27f8a688543a8087951e97b

sweatpotato13 committed:
feat: 🎸 remove tightly coupled dependencies

cocoyoon commented on cumulus/parachains/runtimes/newnal/src/lib.rs:
```diff
@@ -511,6 +511,7 @@ impl pallet_urauth::Config for Runtime {
 	type RuntimeEvent = RuntimeEvent;
+	type Assets = Assets;
 	type MaxPurchaseQuantity = MaxPurchaseQuantity;
 	type TotalFeeRatio = TotalFeeRatio;
 	type MinPlatformFeeRatio = MinPlatformFeeRatio;
```
해당 값들은 트랜잭션으로 수정할 수 있게 팔렛의 configuration 으로 관리하는게 어떨까 합니다. Origin 을 `type Admin` 으로 하고요.

예시,
```rust
struct Configuration { total_fee_ratio, min_plaform, ..};
type PalletConfig = Value<Configuration>; 
```

sweatpotato13 replied:
이 값을 변경할 수 있게 된다면 기존에 존재하는 Contract Detail 또한 스토리지에서 모두 조회해 vaildation을 체크해야 합니다.

사용자의 수수료를 조절하는것이 아닌 플랫폼의 수수료를 조정하는 일이 빈번하게 발생할 것 같지도 않아서 굳이 이를 수정할 수 있도록 하는것이 비효율적이라고 판단 하였습니다.

cocoyoon replied:
네 근데 parameter 변경해줄때마다 런타임 업그레이드를 진행해줘야하는 번거로움이 있지 않을까 해서요

sweatpotato13 replied:
extrinsic으로 변경 시 변경 하고자 하는 값이 기존 Contract의 validation을 깨뜨리는 값이면 변경이 불가능하게 막는 방법으로 구현 하면 좋을 것 같습니다.

sweatpotato13 replied:
생각해보니 fee ratio가 Contract Detail에 포함이 안되어있어 체크할 게 없네요
Extrinsic으로 변경하게 해도 될 것 같습니다.

sweatpotato13 committed:
fix: 🐛 fix hard-coded value

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -44,79 +54,133 @@ pub mod pallet {
+	pub(super) type NextContractId<T: Config> = StorageValue<_, ContractId, ValueQuery>;
+
+	// The Data Trade Records
+	#[pallet::storage]
+	#[pallet::getter(fn get_data_trade_records)]
+	pub type DataTradeRecords<T: Config> =
+		StorageDoubleMap<_, Twox64Concat, ContractId, Twox64Concat, T::AccountId, ()>;
```
`()` 타입 아직 구현안된거죠?

sweatpotato13 replied:
구현 안된게 아니고 아무 데이터도 안들어가 있는게 맞습니다

cocoyoon replied:
`value` 값으로 아무 타입도 안들어갈 수 있나요? 정확히 어떤 의도인가요

sweatpotato13 replied:
특정 ContractID를 가진 Contract를 통해 AccountID가 교환을 진행했는지 여부만을 판단합니다.

이전에 싱크에서 StorageMap, StorageDoubleMap 어떤걸로 할지 얘기가 나왔었는데 그때 DoubleMap으로 하는게 좋을 것 같다고 결론이 나온걸로 알고 있습니다.

cocoyoon replied:
의도가 파악이 안됐습니다ㅠ. 정확히 `DataTradeRecord` 는 어떤 용도인가요? 코드에는 `insert` 부분만 보이고 따로 verify 하는 부분은 안보여서요

energyGiver replied:
data owner가 여러번 판매하지 못하게 막는 용도입니다. DataTradeRecords::<T>::contains_key 체크 하는 if문에 else로 return Err(Error::<T>::AlreadyTraded.into()) 추가 필요해보입니다!

sweatpotato13 replied:
> data owner가 여러번 판매하지 못하게 막는 용도입니다. DataTradeRecords::::contains_key 체크 하는 if문에 else로 return Err(Error::::AlreadyTraded.into()) 추가 필요해보입니다!

한번 데이터 판매하고 종료하는것이 아닌 Contract가 종료될 때 까지 계속 업데이트 내용을 제공받아야 하는 경우가 있을 수 있어 해당 로직은 넣지 않은 상태입니다.

cocoyoon replied:
`()` 타입에 대해 스토리지를 사용해야할지가 고민이네요

cocoyoon replied:
해당 Contract 의 `Account` 를 체크하는 용도라면 `Detail` 구조체 참고해도 되지 않을까 싶습니다

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -44,79 +54,133 @@ pub mod pallet {
+	#[pallet::storage]
+	#[pallet::getter(fn get_contract_status)]
+	pub(super) type ContractStatus<T: Config> =
+		StorageMap<_, Twox64Concat, ContractId, Vec<(T::AccountId, SignStatus)>, OptionQuery>;
```
`Bounded` 있어야 하지 않나요?

sweatpotato13 replied:
최대 2명이라 length limit 2로 걸어두면 될 것 같습니다.

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +441,371 @@ where
+	) -> DispatchResult {
+		let agency = detail.clone().agency;
+		ensure!(maybe_agency == agency, Error::<T>::InvalidAgency);
+		ensure!(detail.clone().effective_at < detail.clone().expired_at, Error::<T>::InvalidPeriod);
```
clone() 이 너무 많이 보이는데 이렇게 하는 것보다 field deconstruct 해서 사용하는게 좋아보입니다

cocoyoon replied:
`maybe_agency == agency` 를 체크하는 것보다 메소드를 두어 caller 를 아예 DelegateContract.agent(who) 로 두는게 나아보입니다
`agency` 필드는 option 타입으로 두고요

cocoyoon replied:
`effective_at` 같은 경우도 처음에 해당 컨트랙트가 생성될때 current_block_num 이나 커스텀하게 두고 `until` 같은거나 글로벌 config `T::Valid` 를 두어 expired_at = effective_at + until 로 두는게 좋아보여요

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +441,371 @@ where
+		ensure!(maybe_agency == agency, Error::<T>::InvalidAgency);
+		ensure!(detail.clone().effective_at < detail.clone().expired_at, Error::<T>::InvalidPeriod);
+
+		let contract_id = NextContractId::<T>::get();
+		NextContractId::<T>::try_mutate(|c| -> DispatchResult {
+			*c = c.checked_add(1).ok_or(Error::<T>::Overflow)?;
+			Ok(())
+		})?;
```
try_mutate 의 반환값으로 현재 contract_id 받아올 수 있습니다

sweatpotato13 replied:
시도해 봤을땐 안되는거같은데 어떤식으로 받아오나요?

sweatpotato13 replied:
URAuth도 위 코드와 동일한 방법으로 사용하고 있고, 레포지토리 내 다른 try_mutate 사용하는 부분을 찾아봐도 일반 value를 return 받아 사용하는 부분은 없는걸로 파악됩니다.

cocoyoon replied:
`try_mutate` 의 반환 타입을 `DispatchResult` 가 아닌 `Result<ContractId, DispatchError>` 로 처리하면 됩니다

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +441,371 @@ where
+		let data_owner = detail.clone().data_owner;
+		let agency = maybe_agency;
+
+		DataDelegateContracts::<T>::insert(contract_id, detail.clone());
+
+		let mut data_owner_list = DataDelegateContractList::<T>::get(&data_owner);
+		data_owner_list.push(contract_id);
+		DataDelegateContractList::<T>::insert(&data_owner, data_owner_list);
```
더 좋은 방법 없을까요..

sweatpotato13 replied:
```rust
DataDelegateContractList::<T>::mutate(&data_owner, |list| {
    list.push(contract_id);
});

DataDelegateContractList::<T>::mutate(&agency, |list| {
    list.push(contract_id);
});
```
이렇게 처리하는 방법이 더 깔끔할 것 같습니다.

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +441,371 @@ where
+
+		let contract_status =
+			vec![(agency.clone(), SignStatus::Signed), (data_owner, SignStatus::Unsigned)];
+		ContractStatus::<T>::insert(contract_id, contract_status.clone());
```
clone 하신 이유가 있나요?

sweatpotato13 replied:
L499 agency는 이후 이벤트 발생 시 사용하므로 clone해 주는것이 맞고 
L500은 contract_status는 clone이 없어야 하는게 맞습니다.

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +441,371 @@ where
+			ContractStatus::<T>::get(contract_id).ok_or(Error::<T>::ContractStatusNotExist)?;
+
+		let mut is_signed = false;
+		for (owner, signed) in status.iter_mut() {
+			if owner == &maybe_owner {
+				*signed = SignStatus::Signed;
+				is_signed = true;
+				break;
+			}
+		}
```
`owner` 를 찾는거라면 `Vec` 보다는 `BTree` 가 나아보입니다. 만약에 해당 키값이 없다면 delegation 이 안된거니 에러를 반환하거나 no-op 처리 하는 식으로요

cocoyoon replied:
다시 insert 해주느것보다 `try_mutate` 로 한번에 가시죠~

cocoyoon commented on substrate/frame/data-market/src/types.rs:
```diff
@@ -50,37 +54,50 @@ impl<BoundedString> VerificationProof<BoundedString> {
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Hash, Debug))]
+pub struct DataDelegateContractDetail<AccountId, BlockNumber, AnyText> {
+	pub data_owner: AccountId,
+	pub data_owner_info: AnyText,
+	pub agency: AccountId,
+	pub agency_info: AnyText,
+	pub data_owner_minimum_fee_ratio: u32,
+	pub deligated_data: AnyText,
+	pub effective_at: BlockNumber,
+	pub expired_at: BlockNumber,
+}
```
`AnyText` 값들은 런타임에서 필요한게 아니라면 따로 관리하는게 맞는거 같습니다

cocoyoon replied:
이 부분은 아직 수정이 안된것으로 보이는데 작업중이신가요?

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +441,371 @@ where
+		>,
+		is_agency_exist: bool,
+	) -> DispatchResult {
+		ensure!(maybe_buyer == detail.data_buyer, Error::<T>::InvalidBuyer);
+		ensure!(detail.clone().effective_at < detail.clone().expired_at, Error::<T>::InvalidPeriod);
```
이쪽도 위와 동일한 리뷰 드립니다

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +441,371 @@ where
+
+		let mut contract_status = vec![(data_buyer.clone(), SignStatus::Signed)];
+
+		if is_agency_exist {
+			ensure!(detail.agency.is_some(), Error::<T>::InvalidAgency);
```
`is_agency_exist` 하나에만 의존하는 것이 아닌 Agency 는 등록된 계정(e.g 거버넌스에 의해 등록)만 할 수있게 2차 방어라인도 구축하는 방향도 좋아보입니다

sweatpotato13 replied:
Agency를 굳이 거버넌스에 의해 등록하기보다 이전에 싱크 미팅에서 얘기가 나온것 처럼 본인 스스로 등록하여 Agency라는 것을 확인할 수 있게 하는 콜이 필요할 것 같습니다.

sweatpotato13 replied:
Commit 726ba5cdff03ee01261f64dc92080e4636f6839b

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +441,371 @@ where
+		if is_agency_exist {
+			ensure!(detail.agency.is_some(), Error::<T>::InvalidAgency);
+			ensure!(detail.data_verifier.is_none(), Error::<T>::InvalidVerifier);
+			let agency = detail.agency.clone().unwrap();
```
물론 위에 .is_some() 통과햇다고는 하지만 `ok_or()` 로 해주시죠 ㅎ

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +441,371 @@ where
+			ensure!(detail.agency.is_some(), Error::<T>::InvalidAgency);
+			ensure!(detail.data_verifier.is_none(), Error::<T>::InvalidVerifier);
+			let agency = detail.agency.clone().unwrap();
+			let mut agency_list = DataPurchaseContractList::<T>::get(&agency);
+			agency_list.push(contract_id);
+			DataPurchaseContractList::<T>::insert(&agency, agency_list);
```
이런 동작들이 반복적으로 보이는데 메소드로 묶어주시죠~

sweatpotato13 replied:
```rust
DataDelegateContractList::<T>::mutate(&data_owner, |list| {
    list.push(contract_id);
});
```
위와 같은 방식으로 수정하였습니다.

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +441,371 @@ where
+
+		ContractStatus::<T>::insert(contract_id, contract_status);
+
+		{
+			let escrow_account = Self::get_escrow_account();
+			Self::transfer_escrow(
+				TransferFrom::Origin(data_buyer.clone()),
+				escrow_account,
+				detail.system_token_id,
+				detail.deposit.into(),
+			)?;
+		}
```
nested 하게 하신 이유가 있나요?

sweatpotato13 replied:
기존 코드를 그대로 재사용 했는데, 지워도 될 것 같네요

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +441,371 @@ where
+		Ok(())
+	}
+
+	fn do_sign_purchase_contract(
```
`do_sign_delegate_contract` 랑 로직이 비슷해 보이는데 하나로 합칠 수 있지 않아요?

sweatpotato13 replied:
delegate와 purchase가 유사하긴 하지만 접근하는 storage와 구조가 조금씩 달라 이를 합친다 해도 결국 일부 로직은 나누어 구분해야 하기 때문에 더 코드가 복잡해 질 것이라 판단하였습니다.

sweatpotato13 committed:
fix: 🐛 add missing deposit update

sweatpotato13 committed:
refactor: 💡 refactor data-market pallet

sweatpotato13 committed:
feat: 🎸 add register agency extrinsic in data-market pallet

sweatpotato13 committed:
refactor: 💡 fix ContractSigner type

BTreeMap -> BoundedBTreeMap

sweatpotato13 committed:
fix: 🐛 add equality check

add equality check for avoiding insert same value in
DataPurchaseContractList & DataDelegateContractList

sweatpotato13 committed:
feat: 🎸 add admin origin in pallet-data-market

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -338,78 +583,47 @@ where
 		data_issuer_fee_ratio: u32,
-	) -> (u128, u128, u128) {
-		let platform_fee_ratio = 10_000 - data_owner_fee_ratio - data_issuer_fee_ratio;
+		agency_fee_ratio: u32,
```
주석으로 공식같이 적어주시면 좋을거같습니다

sweatpotato13 commented:
@cocoyoon 
```rust
pub struct DataDelegateContractDetail<AccountId, BlockNumber> {
	pub data_owner: AccountId,
	pub data_owner_info: AnyText,
	pub agency: AccountId,
	pub agency_info: AnyText,
	pub data_owner_minimum_fee_ratio: u32,
	pub deligated_data: AnyText,
	pub effective_at: BlockNumber,
	pub expired_at: BlockNumber,
}
```
이렇게 변경 되어있는 상태입니다. 
커밋은 들어간거같은데 이상하네요 ㅠ

sweatpotato13 committed:
feat: 🎸 add set config call in data-market

cocoyoon commented on substrate/frame/data-market/src/types.rs:
```diff
@@ -50,37 +58,78 @@ impl<BoundedString> VerificationProof<BoundedString> {
 #[cfg_attr(feature = "std", derive(Hash, Debug))]
-pub struct DataPurchaseRegisterDetails<AccountId, BlockNumber, Balance, AnyText> {
+pub struct DataPurchaseContractDetail<AccountId, BlockNumber, Balance> {
 	pub data_buyer: AccountId,
-	pub data_buyer_info: DataBuyerInfo<AnyText>,
+	pub data_buyer_info: AnyText,
+	pub data_verifier: Option<AccountId>,
+	pub effective_at: BlockNumber,
+	pub expired_at: BlockNumber,
+	pub data_purchase_info: DataPurchaseInfo<AnyText>,
+	pub system_token_id: u32,
+	pub agency: Option<AccountId>,
+	pub agency_info: Option<AnyText>,
+	pub deposit: Balance,
+}
```
제 의견은 AnyText 를 해당 struct 에서 빼는게 좋다고 한거였습니다. 근데 사실상 해당 `info` 는 `identity` 팔렛을 통해 해결할 수 있을거 같은데 불필요하지 않을까 싶습니다.

sweatpotato13 replied:
`identity` 팔렛에 담기는 내용은 조금 더 일반적인 내용이고 `data-market` 에 담기는 내용이 data-market에 좀 더 연관되어 있는 내용이 담길 것이라 생각합니다. 

또한 이 부분을 유지한다 해도 `identity` 팔렛 또한 동시에 사용이 가능하기 때문에 굳이 삭제할 필요까진 없을 것 같다고 생각합니다.

sweatpotato13 replied:
그리고 `AnyText` 는 
```
pub type AnyText = BoundedVec<u8, ConstU32<MAX_TEXT_SIZE>>;
```

단순히 BoundedVec을 정의해놓은 타입인데 이를 굳이 struct에서 제거할 필요가 있나요?

cocoyoon replied:
음 개인적으로 런타임 로직에서 조회할 값도 아닌데 같은 구조체에 넣는게 비이상적이라 생각합니다. `Asset` 같은 경우에도 `AssetDetail`(온체인용) 과 `AssetMetadata` (오프체인용) 으로 빼둔건 이유가 있을거라 생각됩니다

sweatpotato13 replied:
info는 사실상 온체인 오프체인 어디에서도 사용하지 않을 값이라 빼도 상관 없을 것 같습니다.

`deligated_data` 혹은 `data_purchase_info` 같은 필드는 오프체인에서 참고해서 로직에 사용할 값이라고 생각되는데
이를 다른 스토리지에 분리해두면 불필요하게 2번 조회해야 하는 불편함이 생길 수 있을거같은데 이는 유지하는 것이 어떨까 합니다.

cocoyoon commented on substrate/frame/data-market/src/types.rs:
```diff
@@ -50,37 +58,78 @@ impl<BoundedString> VerificationProof<BoundedString> {
+#[derive(Encode, Decode, Clone, Copy, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Hash))]
+pub enum SignStatus {
+	Unsigned,
+	Signed,
```
```rust
pub enum SignStatus {
    Signed(bitmap(00)),
    Terminate(BlockNumber)
}
```
`Signed` 상태는 양자간 합의를 나타내고 `Terminate` 은 언제 하고 싶은지를 나타낼 수 있는 방법으로 할 수 있을거 같은데 어떠신가요?

cocoyoon replied:
아 근데 이런식으로하면 수정할게 많겠군요

sweatpotato13 replied:
`Terminate(BlockNumber)` 는 사실 detail에 포함되어있는 `expired_at`과 중복되는 정보가 될 수 있을 것 같고

`Signed(bitmap(00))` 라고 표시를 하면 어떤 사용자가 Signed 상태인지 명확하지 않을 것 같습니다.

energyGiver commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +682,429 @@ where
+		TradeCountForContract::<T>::insert(contract_id, trade_count);
+		if !DataTradeRecords::<T>::contains_key(contract_id, &data_owner) {
+			DataTradeRecords::<T>::insert(contract_id, &data_owner, ());
+		}
```
contains_key인 경우 error return해주는 구문 추가 필요합니다!

energyGiver commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +682,429 @@ where
+			.map_err(|_| Error::<T>::ExceedContractSigner)?;
+
+		if is_agency_exist {
+			ensure!(agency.is_some(), Error::<T>::InvalidAgency);
```
let agency = agency.ok_or(Error::<T>::InvalidAgency)?;에서 이미 agency option 여부가 체크되어 중복 체크된 걸로 보입니다!

energyGiver commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +682,429 @@ where
+
+		if is_agency_exist {
+			ensure!(agency.is_some(), Error::<T>::InvalidAgency);
+			ensure!(data_verifier.is_none(), Error::<T>::InvalidVerifier);
```
agency 유뮤 상관없이 이쪽에서도 data_verifier.is_some()으로 바뀌어야 할 것 같습니다! if와 else에서 동일한 ensure code이니 외부로 빼도 좋아보입니다.

sweatpotato13 replied:
agency가 있는 경우 data_verifier는 agency가 제시해야 하므로 make하는 시점엔 none이 되는게 맞습니다. 
그렇게 되면 is_agency_exist가 true, false일때 data_verifier는 각각 none, some이 되어야 해 로직이 달라집니다.

energyGiver commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +682,429 @@ where
+			list.push(contract_id);
+		});
+
+		ContractStatus::<T>::insert(contract_id, contract_status);
```
agency 없이 purchase contract 실행할 때, ContractStatus 기록이 필요한가요? 아니라면 아래 코드들은 if is_agency_exist 안에 포함시켜도 좋을 것 같아요.

let mut contract_status: ContractSigner<T> = BoundedBTreeMap::new();
		contract_status
			.try_insert(data_buyer.clone(), SignStatus::Signed)
			.map_err(|_| Error::<T>::ExceedContractSigner)?;

ContractStatus::<T>::insert(contract_id, contract_status);

sweatpotato13 commented:
@cocoyoon 
`DataTradeRecord` https://github.com/InfraBlockchain/infrablockchain-substrate/pull/51#discussion_r1492278814

energyGiver commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -28,266 +35,531 @@ pub mod pallet {
-			data_verifier: T::AccountId,
-			purchase_deadline: BlockNumberFor<T>,
-			system_token_asset_id: u32,
-			quantity: Quantity,
```
make_purchase_contract()에 quantity와 price_per_data가 빠지고 execute_trade_data에 추가되었는데 의도하신 바가 있나요? 구매자가 contract 올리는 시점에 수량과 가격이 정해져야 하는 거 아닌가해서요. 같이 얘기해보면 좋을 것 같습니다!

sweatpotato13 committed:
fix: 🐛 add trade record check

sweatpotato13 committed:
feat: 🎸 add deposit balance call in data-market

sweatpotato13 committed:
fix: 🐛 fix trade record check logic

energyGiver reviewed (approved):
👍🏽

sweatpotato13 committed:
Merge branch 'master' into feature/50-data-market-agency

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -28,266 +35,572 @@ pub mod pallet {
+	#[pallet::getter(fn get_trade_count_for_purchase)]
+	pub type TradeCountForContract<T: Config> =
+		StorageMap<_, Twox64Concat, ContractId, Quantity, ValueQuery>;
 
```
해당 정보 `ContractDetail` 구조체에 field 추가해서 사용해도 되지 않아요?

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -28,266 +35,572 @@ pub mod pallet {
+		OptionQuery,
+	>;
+
+	// The Data Delegate Contract List
+	#[pallet::storage]
+	#[pallet::getter(fn get_data_delegate_contract_list)]
+	pub(super) type DataDelegateContractList<T: Config> =
+		StorageMap<_, Twox64Concat, T::AccountId, Vec<ContractId>, ValueQuery>;
+
+	// The Data Purchase Contract List
+	#[pallet::storage]
+	#[pallet::getter(fn get_data_purchase_contract_list)]
+	pub(super) type DataPurchaseContractList<T: Config> =
+		StorageMap<_, Twox64Concat, T::AccountId, Vec<ContractId>, ValueQuery>;
```
요건 따로 관리해야하는 이유가 있나요? `ContractId` 가 겹치나요?

sweatpotato13 replied:
Contract ID는 해당 pallet에서 글로벌하게 사용되고 있습니다.

Contract ID 할당 규칙을 별도로 규정하지 않는이상 글로벌하게 관리되나, Contract Type별로 별도로 ID를 관리하나 
Contract ID만 보고 Delegate인지 Purchase인지 알 수 없기 때문에 특정 Contract ID가 어떤 Contract인지 확인하기 위해서 여러번 조회 하는것을 방지하기 위해서 위와 같이 구현 하였습니다.

cocoyoon replied:
뭔가 구분은 하되 통합해서 관리할 수 있는 방법이 있을거 같은데 어떤게 좋을까요

sweatpotato13 replied:
제일 간단한건 
ContractId << 1 | 0 -> Delegate
ContractId << 1 | 1 -> Purchase 
이런식으로 LSB로 구분하는게 제일 간단할 것 같습니다.

cocoyoon replied:
이건 어떤 방식인가요? `shr` 해서 `bit_or` 하는 건가요?

sweatpotato13 replied:
left shift 1bit 하고 LSB에 1 혹은 0 을 or로 붙이는 방식입니다.

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -28,266 +35,572 @@ pub mod pallet {
-		StorageMap<_, Twox64Concat, PurchaseId, Quantity, ValueQuery>;
+	#[pallet::getter(fn get_agencies)]
+	pub(super) type Agencies<T: Config> =
+		StorageMap<_, Twox64Concat, T::AccountId, AnyText, ValueQuery>;
```
`StorageValue<Vec<_>>` 만으로도 충분해 보이는데 `AnyText` 로 값을 넣어줘야하는 이유가 있나요? 계속 말씀드리는 부분이지만 해당 Account 에 대한 정보는 `Identity` 팔렛을 이용하면 되는거 같은데 해당 모듈에서까지 넣어줄 필요는 없어 보입니다. 해당 모듈에서는 `Account` 가 등록된 Agency 가 맞는지만 확인하면 될거 같습니다. 그 `Account` 에 대한 정보는 알아서 `Identity` 팔렛에서 확인하고요

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -28,266 +35,572 @@ pub mod pallet {
+	#[pallet::storage]
+	#[pallet::getter(fn get_contract_status)]
+	pub(super) type ContractStatus<T: Config> =
+		StorageMap<_, Twox64Concat, ContractId, ContractSigner<T>, OptionQuery>;
```
차라리 해당 `ContractId` 에 해당하는 `ContractDetail` 구조체에 `signed_status` 필드를 추가해서 메소드를 추가 구현하는게 낫지 않을까요

```suggestion
	ContractDetail {
        ..
        signed_status: [(owner, true/false), (agency, true/false)]
    }
```

sweatpotato13 committed:
fix: 🐛 fix wrong logic about DataTradeRecords

sweatpotato13 committed:
feat: 🎸 remove useless storage

sweatpotato13 committed:
style: 💄 apply formatter

cocoyoon commented on substrate/frame/data-market/src/types.rs:
```diff
@@ -50,37 +57,78 @@ impl<BoundedString> VerificationProof<BoundedString> {
+	pub trade_count: Quantity,
+	pub data_trade_record: Vec<AccountId>,
+	pub signed_status: ContractSigner<AccountId>,
+}
```
두 개를 이런식으로 통합할 수 있을거 같은데 어때요?

```suggestion
pub enum ContractType<AssetKind, Balance> {
	Purchase { asset: AssetKind, deposit: Balance },
	Delegate(Balance /* minimum_fee_ratio */ )
}
pub struct ContractParty<Account>((Account, Account));
impl<Account> ContractParty<Account> {
	fn transfer() { .. }
}	
pub struct Contract<Account, BlockNumber, AssetKind, Balance> {
	pub contract_type: ContractType<AssetKind, Balance>,
	pub party: ContractParty<Account>,
	pub exp: BlockNumber,
	pub eff: BlockNumber,
}
```

sweatpotato13 replied:
각각의 Contract에 대해서 로직이나 Struct를 구성하는 데이터가 차이가 있는데 통합 할 필요가 있을까 싶습니다.

이렇게 구성하더라도 결국에 한쪽에만 있는 데이터들은 별도로 명시를 해 주어야 하는데 그러면 현재 나누어져 있는 구조랑 크게 다르지 않은 것 같습니다.

cocoyoon replied:
로직은 `contract_type` 패턴 분리해서 구현한다면 서로 분리된 function 도 통합하여 구현할 수 있을거 같아서요. 필드값이 차이가 나지 않는것으로 보이고 이건 나중 task 로 잡아도 되지 않을까 싶긴합니다..

cocoyoon commented on substrate/frame/data-market/src/types.rs:
```diff
@@ -50,37 +57,78 @@ impl<BoundedString> VerificationProof<BoundedString> {
 	}
```
```suggestion
	VerificationProof(BoundedString)
```

sweatpotato13 replied:
```rust
impl<BoundedString> VerificationProof<BoundedString> {
	pub fn new(hash: BoundedString) -> Self {
		Self { hash }
		VerificationProof(BoundedString)  
}
```

sweatpotato13 replied:
Tuple로 하자는 의견인가요?

cocoyoon replied:
필드가 하나여서 union 타입으로 하는게 어떨까 싶어서요

sweatpotato13 replied:
필드가 하나일때 union으로 하는게 struct로 하는 것에 비해서 이점이 있나요?

sweatpotato13 replied:
substrate에서 union 타입 가능한가요?
Union types are not supported. 라고 나오는데

sweatpotato13 replied:
그리고 오히려 필드가 하나인 경우엔 union과 struct간 성능 차이가 미미할 것으로 보입니다.

sweatpotato13 replied:
사실 VerificationProof에 대해서 struct로 구성할 게 아니고 `BoundedVec` 처리하는게 오히려 더 깔끔할 것 같습니다.

로직에서 사용하는 값이 아닌 Event에서만 사용하는 값이라 이 조치로 충분할 것 같습니다.

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +751,435 @@ where
+				let _ = T::Assets::transfer(
 					system_token_asset_id.into(),
 					&origin,
 					&to,
 					balance,
-					None,
-					f,
-				)
-				.map(|_| ())?;
+					Preservation::Protect,
+				);
 			},
 			TransferFrom::Escrow => {
```
`TransferFrom` enum 을 둔 이유가 궁금합니다. `from` 파라미터만 잘 넣어주면 되는거 아닌가요?

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +751,435 @@ where
 		system_token_asset_id: u32,
 		amount: u128,
```
해당 모듈에 있는 `balance` 타입들 모두 `AssetBalanceOf<T>` 로 맞추면 타입 변환할 필요없지 않을까 싶습니다

sweatpotato13 replied:
단순히 balance만 바꿔준다고 해서 타입 변환이 없어지지는 않을 것 같습니다. weight등 엮여 있는 부분이 많아서 그런 부분도 다 변경해 주어야 하네요

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -418,35 +751,435 @@ where
+		data_verification_proof: VerificationProof<AnyText>,
+	) -> DispatchResult {
+		let mut detail =
+			DataPurchaseContracts::<T>::get(contract_id).ok_or(Error::<T>::ContractNotExist)?;
```
`PurchaseDetail { .. } = *::get()` 로 deconstruct 하는게 좋아보입니다

sweatpotato13 committed:
fix: 🐛 fix inefficiency logic

cocoyoon commented on substrate/frame/data-market/src/lib.rs:
```diff
@@ -28,266 +35,570 @@ pub mod pallet {
+	// The Data Delegate Contract List
+	#[pallet::storage]
+	#[pallet::getter(fn get_data_delegate_contract_list)]
+	pub(super) type DataDelegateContractList<T: Config> =
+		StorageMap<_, Twox64Concat, T::AccountId, Vec<ContractId>, ValueQuery>;
+
+	// The Data Purchase Contract List
 	#[pallet::storage]
-	#[pallet::getter(fn data_trade_records)]
-	pub type DataTradeRecords<T: Config> =
-		StorageDoubleMap<_, Twox64Concat, PurchaseId, Twox64Concat, T::AccountId, ()>;
+	#[pallet::getter(fn get_data_purchase_contract_list)]
+	pub(super) type DataPurchaseContractList<T: Config> =
+		StorageMap<_, Twox64Concat, T::AccountId, Vec<ContractId>, ValueQuery>;
```
`ContractType` 타입이 정의가 되어있으니 `key1 = T::AccountId` `key2 = ContractType` `value = BoundedVec<ContractId, MaxContract>` 하면 되지 않을까 싶네요

cocoyoon commented on substrate/frame/data-market/src/types.rs:
```diff
@@ -3,15 +3,22 @@ use super::*;
 pub type Quantity = u128;
 pub type IssuerWeight = u32;
 
 /// Common size is up to 100 bytes
 pub const MAX_TEXT_SIZE: u32 = 1_000_000;
+pub const MAX_ENTITIES: u32 = 2;
```
해당 값도 `Config` 의 `constant` 로 두죠

sweatpotato13 replied:
이 부분은 로직 수정이 없으면 절대 변하지 않을 값이라서 굳이 지금 Config에 둘 필요가 없을 것 같습니다.

sweatpotato13 replied:
이후 로직 수정이 되서 2명 이상의 유저가 하나의 Contract에 묶일 때 그때 Config에 넣어도 될 것 같습니다.

sweatpotato13 committed:
feat: 🎸 merge contract storages

energyGiver reviewed (approved):
👍
